kolibri/core/auth/management/commands/bulkimportusers.py

```diff
@@ -0,0 +1,587 @@
+    "facility coach": "coach",
+    "class coach": "classroom assignable coach",
+}
+
+# Validators ###
+
+
+def number_range(min, max):
+    """
+    Return a value check function which raises a ValueError if the supplied
+    value is less than `min` or greater than `max`.
+    """
+
+    def checker(v):
+        if int(v) < min or int(v) > max:
+            raise ValueError(v)
+
+    return checker
+
+
+def value_length(length, null=False):
+    """
+    Return a value check function which raises a ValueError if the supplied
+    value has a length greater than 'length'
+    If null is not True raises a ValueError if the supplied value is None.
+    """
+
+    def checker(v):
+        if null and v is None:
+            return checker
+        if len(v) > length:
+            raise ValueError(v)
+
+    return checker
+
+
+def enumeration(*args):
+    """
+    Return a value check function which raises a ValueError if the value (case
+    insensitive) is not in the enumeration of values provided by args.
+    """
+    assert len(args) > 0, "at least one argument is required"
+    if len(args) == 1:
+        # assume the first argument defines the membership
+        members = args[0].lower()
+    else:
+        members = tuple(map(str.lower, args))
+
+    def checker(value):
+        if value.lower() not in members:
+            raise ValueError(value)
+
+    return checker
+
+
+def valid_name(username=True, null=False):
+    """
+    Return a value check function which raises a ValueError if the value has
+    some of the punctuaction chars that are not allowed.
+    If username is False it allows spaces, slashes and hyphens.
+    If null is not True raises a ValueError if the supplied value is None.
+    """
+
+    def checker(v):
+        if null and v is None:
+            return checker
+        has_punc = "/[\s`~!@#$%^&*()\-+={}\[\]\|\\\/:;\"'<>,\.\?]/"  # noqa
+        if not username:
+            has_punc = "/[`~!@#$%^&*()\+={}\[\]\|\\\/:;\"'<>\.\?]/"  # noqa
+        if re.match(has_punc, v):
+            raise ValueError(v)
+
+    return checker
+
+
+class Validator(object):
+    """
+    Class to apply different validation checks on a CSV data reader.
+    """
+
+    def __init__(self, header_names):
+        self._header_names = header_names
+        self._checks = list()
+        self.classrooms = dict()
+        self.coach_classrooms = dict()
+        self.users = dict()
+        self.roles = {r: list() for r in roles_map.values() if r is not None}
+
+    def add_check(self, header_name, check, message):
+        """
+        Add a header check, i.e., check whether the header record is consistent
+        with the expected field names.
+        `header_name` - name of the header for the column to be checked
+        `check`- function to be used as validator of the values in the column
+        `message` - problem message to report if a value is not valid
+        """
+        self._checks.append((header_name, check, message))
```

Comment:
rather than a `message`, perhaps we should instead be creating validation codes? Then on the client side we can map code to internationalized messages?

Alternatively, we could do internationalization on the backend.

Reply:
yes, I thought of it but I also thought this is a mvp and I prefered not to spend time on it now. I prefer to test the behaviour of the script asap. Same thing with the progress_update calculations that are not real yet.

Reply:
okay